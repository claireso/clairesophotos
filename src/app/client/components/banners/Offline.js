import React from 'react'

import useIsOnline from '@common/hooks/isOnline'
import TranslationsContext from '@common/context/Translations'

import Flash from '../Flash'

const BannerOffline = () => {
  const isOnline = useIsOnline()

  if (isOnline) return null

  return (
    <TranslationsContext.Consumer>
      {translations => <Flash>{translations.bannerOffline}</Flash>}
    </TranslationsContext.Consumer>
  )
}

export default BannerOffline
