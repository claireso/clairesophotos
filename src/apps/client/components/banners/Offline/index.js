import React, { useContext } from 'react'

import useIsOnline from '@common/hooks/isOnline'
import TranslationsContext from '@common/context/Translations'

import Flash from '../../Flash'

const BannerOffline = () => {
  const isOnline = useIsOnline()
  const translations = useContext(TranslationsContext)

  if (isOnline) return null

  return <Flash>{translations.bannerOffline}</Flash>
}

export default BannerOffline
