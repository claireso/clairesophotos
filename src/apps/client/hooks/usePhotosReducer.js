import { useReducer } from 'react'

import { getPhotos } from '@client/api'
import photosReducer, { TYPES } from './reducer/photos'

export default (initialState = {}) => {
  const [state, dispatch] = useReducer(photosReducer, initialState)

  return [
    state,
    {
      setLoading: isLoading =>
        dispatch({ type: TYPES.SET_LOADING, isLoading: isLoading }),
      getPhotos: async page => {
        try {
          const response = await getPhotos(page)

          dispatch({ type: TYPES.GET_PHOTOS_SUCCESS, response })
        } catch (err) {
          throw new Error('Can not get photos')
        }
      }
    }
  ]
}
