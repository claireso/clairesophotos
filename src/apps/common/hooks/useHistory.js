import { useEffect } from 'react'

export default ({ onPopstate } = {}) => {
  useEffect(() => {
    onPopstate && window.addEventListener('popstate', onPopstate)

    return () => {
      onPopstate && window.removeEventListener('popstate', onPopstate)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const pushState = (state, title, url) => {
    window.history.pushState(state, title, url)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }

  return {
    pushState
  }
}
