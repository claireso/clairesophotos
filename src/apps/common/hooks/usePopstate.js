import { useEffect } from 'react'

export default fn => {
  useEffect(() => {
    // listen history
    window.addEventListener('popstate', fn)

    return () => {
      window.removeEventListener('popstate', fn)
    }
  }, [fn])
}
