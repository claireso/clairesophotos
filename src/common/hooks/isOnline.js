import { useState, useEffect } from 'react'

export default () => {
  const [isOnline, setIsOnline] = useState(true)

  const handleOnlineStatus = () => setIsOnline(true)

  const handleOfflineStatus = () => setIsOnline(false)

  useEffect(() => {
    setIsOnline(navigator.onLine)

    window.addEventListener('offline', handleOfflineStatus)
    window.addEventListener('online', handleOnlineStatus)

    return () => {
      window.removeEventListener('offline', handleOfflineStatus)
      window.removeEventListener('online', handleOnlineStatus)
    }
  }, [])

  return isOnline
}