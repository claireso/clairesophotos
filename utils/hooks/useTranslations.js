import React from 'react'
import PropTypes from 'prop-types'

const TranslationsContext = React.createContext({})

const TranslationsProvider = ({ children, translations }) => (
  <TranslationsContext.Provider value={translations}>
    {children}
  </TranslationsContext.Provider>
)

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

const useTranslations = () => React.useContext(TranslationsContext)

export { TranslationsProvider, useTranslations }
