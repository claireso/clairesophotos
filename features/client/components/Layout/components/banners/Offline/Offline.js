import React from 'react'

import { useTranslations } from '@utils/hooks/useTranslations'

import Flash from '@components/client/Flash'

import useIsOnline from './hooks/useIsOnline'

const BannerOffline = () => {
  const isOnline = useIsOnline()
  const translations = useTranslations()

  if (isOnline) return null

  return <Flash>{translations.bannerOffline}</Flash>
}

export default BannerOffline
