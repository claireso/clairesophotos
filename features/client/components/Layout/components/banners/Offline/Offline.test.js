import { render } from '@testing-library/react'

import { TranslationsProvider } from '@utils/hooks/useTranslations'
import Offline from './index'

describe('<Offline />', () => {
  const renderBanner = () =>
    render(
      <TranslationsProvider translations={global.__TRANSLATIONS__.client}>
        <Offline />
      </TranslationsProvider>
    )

  test('should not render component', () => {
    const { container } = renderBanner()

    expect(container).toMatchSnapshot()
  })

  test('should render component', () => {
    global.goOffline()

    const { container } = renderBanner()

    expect(container).toMatchSnapshot()

    global.goOnline()
  })
})
