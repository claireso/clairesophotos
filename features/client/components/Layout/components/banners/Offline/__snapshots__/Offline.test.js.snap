// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Offline /> should not render component 1`] = `<div />`;

exports[`<Offline /> should render component 1`] = `
.c0 {
  background: var(--flash-default);
  color: var(--flash-default-text);
  font-size: var(--font-size-normal);
  margin: 0 0 var(--gutter);
  padding: 2rem;
  position: relative;
  text-align: center;
}

<div>
  <div
    class="c0"
    style="opacity: 0;"
  >
    You are currently offline. Connect to the internet for access to the latest photos
  </div>
</div>
`;
